File: schedule_forge/scheduling/rooms.py
"""Allocating rooms to the exams of a system (the module of version 3.0).

Every exam of a date is seated in rooms of its own: a room holds one exam on a
given date, so two exams never overlap in a room, exactly the way two exams of
one study year never overlap on a date. An exam that needs more seats than the
largest free room has is spread over several rooms.

The allocation is done for one exam system at a time, after its dates are fixed.
"""

from ..model.room import RoomAllocation, RoomBooking


class RoomAllocator(object):
    """Seats the exams of an exam system in the rooms of the campus."""

    def __init__(self, rooms, default_students=30):
        #: Smallest room first: an exam is given the smallest room it fits in,
        #: which keeps the large rooms free for the exams that need them.
        self.rooms = sorted(rooms, key=lambda room: (room.capacity, room.name))
        self.default_students = default_students

    @property
    def total_capacity(self):
        return sum(room.capacity for room in self.rooms)

    def students_of(self, exam):
        return exam.course.students_or(self.default_students)

    def allocate(self, system):
        """Return the `RoomAllocation` of a whole exam system."""
        by_date = {}
        for scheduled in system.scheduled_exams:
            by_date.setdefault(scheduled.date, []).append(scheduled)

        bookings = {}
        failures = []
        for date in sorted(by_date):
            self._allocate_day(date, by_date[date], bookings, failures)
        return RoomAllocation(bookings, failures)

    def can_allocate(self, system):
        """True when every exam of the system can be seated."""
        return self.allocate(system).is_complete

    def _allocate_day(self, date, scheduled_exams, bookings, failures):
        # The exam with the most students picks first: it is the one that a
        # late pick would leave without a room large enough.
        ordered = sorted(scheduled_exams,
                         key=lambda item: (-self.students_of(item.exam),
                                           item.exam.course.number))
        free = list(self.rooms)
        for scheduled in ordered:
            needed = self.students_of(scheduled.exam)
            taken = self._take_rooms(free, needed)
            if taken is None:
                failures.append(
                    "%s %s on %s needs %d seats, only %d are free that day"
                    % (scheduled.exam.course.number, scheduled.exam.course.name,
                       date.strftime("%d-%m-%Y"), needed,
                       sum(room.capacity for room in free)))
                continue
            for room in taken:
                free.remove(room)
            bookings[scheduled.exam.key] = RoomBooking(scheduled.exam, date, taken)

    def _take_rooms(self, free, needed):
        """The rooms one exam is given, or None when it cannot be seated."""
        for room in free:  # smallest first: the smallest room that holds it all
            if room.capacity >= needed:
                return [room]
        # No single room is large enough, so fill it with the largest ones.
        taken = []
        seats = 0
        for room in reversed(free):
            taken.append(room)
            seats += room.capacity
            if seats >= needed:
                return taken
        return None

File: web/src/engine/rooms.ts
/**
 * Allocating rooms to the exams of a system (the module of version 3.0).
 *
 * Every exam of a date is seated in rooms of its own: a room holds one exam on
 * a given date, so two exams never overlap in a room, exactly the way two exams
 * of one study year never overlap on a date. An exam that needs more seats than
 * the largest free room has is spread over several rooms.
 */

import { Exam, ExamSystem, Room } from "./model";

export interface RoomBooking {
  exam: Exam;
  date: string;
  rooms: Room[];
}

export interface RoomAllocation {
  bookings: Map<string, RoomBooking>;
  failures: string[];
  isComplete: boolean;
}

export class RoomAllocator {
  /** Smallest room first: an exam takes the smallest room it fits in. */
  private readonly rooms: Room[];

  constructor(rooms: Room[], private readonly defaultStudents = 30) {
    this.rooms = [...rooms].sort(
      (a, b) => a.capacity - b.capacity || a.name.localeCompare(b.name)
    );
  }

  get totalCapacity(): number {
    return this.rooms.reduce((sum, room) => sum + room.capacity, 0);
  }

  studentsOf(exam: Exam): number {
    return exam.course.students ?? this.defaultStudents;
  }

  allocate(system: ExamSystem): RoomAllocation {
    const byDate = new Map<string, ExamSystem>();
    for (const scheduled of system) {
      const list = byDate.get(scheduled.date) ?? [];
      list.push(scheduled);
      byDate.set(scheduled.date, list);
    }

    const bookings = new Map<string, RoomBooking>();
    const failures: string[] = [];
    for (const date of [...byDate.keys()].sort()) {
      this.allocateDay(date, byDate.get(date)!, bookings, failures);
    }
    return { bookings, failures, isComplete: failures.length === 0 };
  }

  private allocateDay(
    date: string,
    scheduledExams: ExamSystem,
    bookings: Map<string, RoomBooking>,
    failures: string[]
  ): void {
    // The exam with the most students picks first: it is the one a late pick
    // would leave without a room large enough.
    const ordered = [...scheduledExams].sort(
      (a, b) =>
        this.studentsOf(b.exam) - this.studentsOf(a.exam) ||
        a.exam.course.number.localeCompare(b.exam.course.number)
    );
    const free = [...this.rooms];
    for (const scheduled of ordered) {
      const needed = this.studentsOf(scheduled.exam);
      const taken = takeRooms(free, needed);
      if (!taken) {
        const seats = free.reduce((sum, room) => sum + room.capacity, 0);
        failures.push(
          `${scheduled.exam.course.number} ${scheduled.exam.course.name} on ` +
            `${date} needs ${needed} seats, only ${seats} are free that day`
        );
        continue;
      }
      for (const room of taken) free.splice(free.indexOf(room), 1);
      bookings.set(scheduled.exam.id, { exam: scheduled.exam, date, rooms: taken });
    }
  }
}

function takeRooms(free: Room[], needed: number): Room[] | null {
  for (const room of free) {
    if (room.capacity >= needed) return [room]; // the smallest room that holds it
  }
  // No single room is large enough, so fill it with the largest ones.
  const taken: Room[] = [];
  let seats = 0;
  for (let index = free.length - 1; index >= 0; index -= 1) {
    taken.push(free[index]);
    seats += free[index].capacity;
    if (seats >= needed) return taken;
  }
  return null;
}
